core/internal/filetransfer/file_transfer_default_test.go

```diff
@@ -105,3 +107,82 @@ func TestDefaultFileTransfer_Upload(t *testing.T) {
 	err = ft.Upload(task)
 	assert.NoError(t, err)
 }
+
+func TestDefaultFileTransfer_UploadNotFound(t *testing.T) {
+	fnfHandler := func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusNotFound)
+	}
+	err := uploadToServerWithHandler(t, fnfHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "404")
+}
+
+func TestDefaultFileTransfer_UploadConnectionClosed(t *testing.T) {
+	closeHandler := func(w http.ResponseWriter, r *http.Request) {
+		hj, ok := w.(http.Hijacker)
+		assert.True(t, ok, "webserver doesn't support hijacking")
+		conn, _, err := hj.Hijack()
+		assert.NoError(t, err, "hijacking error")
+		conn.Close()
+	}
+	err := uploadToServerWithHandler(t, closeHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "giving up after 2 attempt(s)")
+}
+
+func TestDefaultFileTransfer_UploadNoResponse(t *testing.T) {
+	err := uploadToServerWithHandler(t, func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
+	})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "context deadline exceeded")
+}
```

Comment:
500ms will add up quickly; we'd like to keep these tests fast.

This test is just checking the `err != nil` case in `file_transfer_default.go`. You should be able to trigger that codepath by cancelling the context instead, which won't require sleeping. https://pkg.go.dev/context#WithCancel

Reply:
Could you explain what you mean? Empirically, that doesn't seem to elicit the same behavior. Neither calling `cancel()` nor sending `<-ctx.Done()` causes a context deadline exceeded error.

Reply:
The "context deadline exceeded" error comes from the standard `http` library, not from the code being tested. This test checks (/ breaks if you could change the fact) that:

* `http` respects context deadlines and returns an error
* `retryablehttp` passes that error through
* our code passes that error through (rather than ignoring it)

Tests can assume that dependencies are correct, so only the last point is useful to test, and there's a way to trigger it without a sleep.

Reply:
I think I don't understand the mechanics of what you're suggesting. I've tried
```
_, cancel := context.WithCancel(r.Context())
cancel()
```
which returns a nil error, and
```
ctx, _ := context.WithCancel(r.Context())
<-ctx.Done()
```
which causes it to halt and never return. How should I be canceling the context?

Reply:
No, you're right, I realize we don't set a Context at all. I got it to work with:

```go
// Add Context field to Task:
type Task struct {
	// An optional context for the task.
	//
	// This can be used to cancel the file upload or download if it is no
	// longer needed.
	Context context.Context

	...
}

// Use Context in Upload():
if task.Context != nil {
	req = req.WithContext(task.Context)
}

// Test:
ctx, cancel := context.WithCancel(context.Background())
cancel()

err = ft.Upload(&filetransfer.Task{
	Context: ctx,
	Type:    filetransfer.UploadTask,
	Path:    tempFile.Name(),
	Url:     server.URL,
})
```

Another valid way to make `retryablehttp.Client` to return a non-nil err is to give it a custom RoundTripper that returns a non-nil err. That wouldn't require updating any non-test code:

```go
// Right above the test declaration:
type failingTransport struct {
	msg string
}

func (t *failingTransport) RoundTrip(_ *http.Request) (*http.Response, error) {
	return nil, fmt.Errorf(t.msg)
}

// In the test body:
client := impatientClient()
client.HTTPClient.Transport = &failingTransport{msg: "test test test"}
ft := filetransfer.NewDefaultFileTransfer(observability.NewNoOpLogger(), client)

...

assert.Contains(t, err.Error(), "test test test")
```

Reply:
Sorry I backburnered this, getting back to it now.

I still might be a little confused; I used your first solution (having context on the task seems useful I guess, I'll bow to your opinion there), but that results in a different error and even more relevantly, different retry behavior. I kept the "context cancellation" test in and reduced the timeout to 50ms–if that still doesn't work for you I'm happy to just remove that test entirely.

Reply:
Another reason to avoid sleeping in tests, other than performance, is flakiness. Sleeping threads aren't guaranteed to wake up immediately after the requested duration, and there are no ordering guarantees here, so there is a race condition in this test:

1. The http library waits for the deadline to pass
2. The test's request handler sleeps

These could finish in any order regardless of their actual durations.

---

Here are our test timings on `main` right now:

```
ok  	github.com/wandb/wandb/core/internal/api	0.478s
ok  	github.com/wandb/wandb/core/internal/clients	0.252s
ok  	github.com/wandb/wandb/core/internal/data_types	0.351s
ok  	github.com/wandb/wandb/core/internal/debounce	0.499s
ok  	github.com/wandb/wandb/core/internal/filetransfer	0.250s
ok  	github.com/wandb/wandb/core/internal/gqlmock	0.489s
ok  	github.com/wandb/wandb/core/internal/runconfig	0.475s
ok  	github.com/wandb/wandb/core/internal/runfiles	1.063s
ok  	github.com/wandb/wandb/core/internal/shared	0.591s
ok  	github.com/wandb/wandb/core/internal/watcher	0.839s
ok  	github.com/wandb/wandb/core/pkg/artifacts	0.620s
ok  	github.com/wandb/wandb/core/pkg/filestream	4.707s
ok  	github.com/wandb/wandb/core/pkg/launch	0.637s
ok  	github.com/wandb/wandb/core/pkg/leveldb	0.763s
ok  	github.com/wandb/wandb/core/pkg/observability	0.653s
ok  	github.com/wandb/wandb/core/pkg/server	0.619s
ok  	github.com/wandb/wandb/core/pkg/utils	0.560s
```

`filestream` is unfortunately quite long already, taking up 5 whole seconds compared to other tests. The other tests average out to around 500ms per test. 50ms is significantly better than 500ms, but it is still around 10% of a more reasonably sized test.

But all that aside, my point is that the check `assert.Contains(t, err.Error(), "context deadline exceeded")` is testing a dependency, not the code under test.

Reply:
like I said, happy to remove it entirely

Reply:
Oops, I missed that! Sorry for the back and forth, I admit this is only a minor thing. I do think it's better to remove that particular test.